emu/src/cpu/data_processing.rs

```diff
@@ -276,27 +276,41 @@ impl Arm7tdmi {
 
                 let current_mode = self.cpsr.mode();
 
-                let psr = match p {
-                    false => &mut self.cpsr,
-                    true => self.get_spsr_as_ref_mut(),
-                };
+                if matches!(self.cpsr.mode(), Mode::System | Mode::User) && p {
+                    panic!("Can't access SPSR in System/User mode")
+                }
 
-                // Setting flags
-                psr.set_sign_flag(rm.get_bit(31));
-                psr.set_zero_flag(rm.get_bit(30));
-                psr.set_carry_flag(rm.get_bit(29));
-                psr.set_overflow_flag(rm.get_bit(28));
-
-                // In User mode we can only set the flags so we don't touch the Mode bits
-                if current_mode != Mode::User {
-                    psr.set_irq_disable(rm.get_bit(7));
-                    psr.set_fiq_disable(rm.get_bit(6));
-
-                    // Documentation says that software should never touch T (state) bit
-                    // Should we set it? I guess software are written in order to not switch this bit
-                    // but who knows?
-                    // psr.set_state_bit(rm.get_bit(5));
-                    psr.set_mode_raw(rm.get_bits(0..=4));
+                {
+                    let psr = match p {
+                        false => &mut self.cpsr,
+                        true => &mut self.spsr,
+                    };
+                    // Setting flags
+                    psr.set_sign_flag(rm.get_bit(31));
+                    psr.set_zero_flag(rm.get_bit(30));
+                    psr.set_carry_flag(rm.get_bit(29));
+                    psr.set_overflow_flag(rm.get_bit(28));
```

Comment:
You can use `pub fn set_flags(&mut self, op_result: ArithmeticOpResult)` for this purpose

Reply:
But then we shold construct a "dummy" `ArithmeticOpResult`?

Reply:
Yes is a borderline case

Reply:
But I don't know it would look weird to have an `ArithmeticOpResult` here without any arithmetic operation 😅